rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+		os << ">";
+		os << "</binary_nodes_and_edges>";
+	}
+	if((*this).has_connection_boxes){
+		os << "<connection_boxes";
+		if((bool)(*this).connection_boxes.x_dim)
+			os << " x_dim=\"" << (*this).connection_boxes.x_dim << "\"";
+		if((bool)(*this).connection_boxes.y_dim)
+			os << " y_dim=\"" << (*this).connection_boxes.y_dim << "\"";
+		if((bool)(*this).connection_boxes.num_boxes)
+			os << " num_boxes=\"" << (*this).connection_boxes.num_boxes << "\"";
+		os << ">";
+		for(auto &connection_box: (*this).connection_boxes.connection_boxes){
+			os << "<connection_box";
+			if((bool)connection_box.id)
+				os << " id=\"" << connection_box.id << "\"";
+			if((bool)connection_box.name)
+				os << " name=\"" << connection_box.name << "\"";
+			os << ">";
+			os << "</connection_box>";
+		}
+		os << "</connection_boxes>";
+	}
+	os << "<channels>";
+	os << "<channel";
+	os << " chan_width_max=\"" << (*this).channels.channel.chan_width_max << "\"";
+	os << " x_min=\"" << (*this).channels.channel.x_min << "\"";
+	os << " y_min=\"" << (*this).channels.channel.y_min << "\"";
+	os << " x_max=\"" << (*this).channels.channel.x_max << "\"";
+	os << " y_max=\"" << (*this).channels.channel.y_max << "\"";
+	os << ">";
+	os << "</channel>";
+	for(auto &x_list: (*this).channels.x_lists){
+		os << "<x_list";
+		os << " index=\"" << x_list.index << "\"";
+		os << " info=\"" << x_list.info << "\"";
+		os << ">";
+		os << "</x_list>";
+	}
+	for(auto &y_list: (*this).channels.y_lists){
+		os << "<y_list";
+		os << " index=\"" << y_list.index << "\"";
+		os << " info=\"" << y_list.info << "\"";
+		os << ">";
+		os << "</y_list>";
+	}
+	os << "</channels>";
+	os << "<switches>";
+	for(auto &switch_: (*this).switches.switches){
+		os << "<switch";
+		os << " id=\"" << switch_.id << "\"";
+		os << " name=\"" << switch_.name << "\"";
+		if((bool)switch_.type)
+			os << " type=\"" << lookup_switch_type[(int)switch_.type] << "\"";
+		os << ">";
+		if(switch_.has_timing){
+			os << "<timing";
+			if((bool)switch_.timing.R)
+				os << " R=\"" << switch_.timing.R << "\"";
+			if((bool)switch_.timing.Cin)
+				os << " Cin=\"" << switch_.timing.Cin << "\"";
+			if((bool)switch_.timing.Cinternal)
+				os << " Cinternal=\"" << switch_.timing.Cinternal << "\"";
+			if((bool)switch_.timing.Cout)
+				os << " Cout=\"" << switch_.timing.Cout << "\"";
+			if((bool)switch_.timing.Tdel)
+				os << " Tdel=\"" << switch_.timing.Tdel << "\"";
+			os << ">";
+			os << "</timing>";
+		}
+		os << "<sizing";
+		os << " mux_trans_size=\"" << switch_.sizing.mux_trans_size << "\"";
+		os << " buf_size=\"" << switch_.sizing.buf_size << "\"";
+		os << ">";
+		os << "</sizing>";
+		os << "</switch>";
+	}
+	os << "</switches>";
+	os << "<segments>";
+	for(auto &segment: (*this).segments.segments){
+		os << "<segment";
+		os << " id=\"" << segment.id << "\"";
+		os << " name=\"" << segment.name << "\"";
+		os << ">";
+		if(segment.has_timing){
+			os << "<timing";
+			if((bool)segment.timing.R_per_meter)
```

Comment:
Is a zero value for `R_per_meter` valid?

Reply:
No - we are losing some of the file here. Probably a `<timing R=0 C=0/>` wouldn't make it to the other side.

However, this check is only generated for optional attributes.
```xsd
  <xs:complexType name="segment_timing">
    <xs:attribute name="R_per_meter" type="xs:float"/><!-- no use="required" here -->
    <xs:attribute name="C_per_meter" type="xs:float"/>
  </xs:complexType>
```
I don't think it is too wrong to assume a value of 0 for non-existing optional attributes? To _really_ correct this, we would have to add a bool `has_R_per_meter` for every attribute.

Reply:
Does the XML schema support default value description?

Reply:
Yes, for the optional attributes. If the optional attribute has a nonzero default value, we print it in all cases: https://github.com/duck2/uxsdcxx/blob/master/uxsdcxx.py#L756